### PR TITLE
Issue #4465: Rename LocalizedMessage to Violation

### LINKED ISSUE #4465: Rename LocalizedMessage to Violation
Based on https://github.com/checkstyle/checkstyle/issues/4419, AST node which caused a violation should be a part of LocalizedMessage. LocalizedMessage will become the class which stores all information about a violation. LocalizedMessage is an inappropriate name for the class. From my point of view, we need to rename the class to CheckstyleViolation.

### PR DESCRIPTION
**Pull request by AkMo3**
Closes #4465 

```
$ grep -r "localizedMessage"
<--No Output-> 

$ grep -r "LocalizedMessage"
src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java:            final String causeMessage = ex.getCause().getLocalizedMessage();
src/test/java/com/puppycrawl/tools/checkstyle/checks/UniquePropertiesCheckTest.java:            return ex.getLocalizedMessage();
src/test/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheckTest.java:            return ex.getLocalizedMessage();
src/main/java/com/puppycrawl/tools/checkstyle/checks/UniquePropertiesCheck.java:                    ex.getLocalizedMessage());
src/main/java/com/puppycrawl/tools/checkstyle/checks/OrderedPropertiesCheck.java:            log(1, MSG_IO_EXCEPTION_KEY, file.getPath(), ex.getLocalizedMessage());

```
These `LocalizedMessaged` are of Exception class localised messaged

### CODE REVIEW

**Comment by AkMo3** [comment]
@romani @pbludov @strkkk Sir, sonar giving these errors https://sonarcloud.io/dashboard?id=org.checkstyle%3Acheckstyle&pullRequest=9761, please tell how to fix them? Do I have to change the code too?

File: src/main/java/com/puppycrawl/tools/checkstyle/DefaultLogger.java [hunk 1]
```diff
@@ -159,29 +159,29 @@ public void addError(AuditEvent event) {
...
             errorWriter.println(addExceptionMessage.getMessage());
             throwable.printStackTrace(errorWriter);
         }
     }
 
     @Override
     public void auditStarted(AuditEvent event) {
-        final LocalizedMessage auditStartMessage = new LocalizedMessage(1,
+        final Violation auditStartMessage = new Violation(1,
             Definitions.CHECKSTYLE_BUNDLE, AUDIT_STARTED_MESSAGE, null, null,
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/DefaultLogger.java:174]: @romani Here is an example of this not being a violation.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/DefaultLogger.java:174]: yes, good catch!!

Looks like we need to keep LocalizedMessage class for cases like this.
So we need to 2 classes LocalizedMessage for Loggers and Violation for Checks.
LocalizedMessage should have no fields for column, line, ... tokenType, ... .
we can make separate issue on this.

Right now this Violation class might looks a bit weird in context of Loggers, but look at arguments they mimic empty violation. After class is created we immediately convert to String by
https://github.com/checkstyle/checkstyle/blob/f868b02a37191ca662bb919f44353b55897793b0/src/main/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessage.java#L454

So I am ok with such update (separate issue to create LocalizedMessage class that do a localization for string values).

Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/DefaultLogger.java:174]: Sir, can you make a separate issue, I can start working on it right away.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/DefaultLogger.java:174]: https://github.com/checkstyle/checkstyle/issues/9794

**Comment by AkMo3** [comment]
@romani Sir, can we solve #9794 and #4465 in this single PR

**Comment by romani** [comment]
> can we solve #9794 and #4465 in this single PR

it is better to do separately.
it is ok in this PR to let Loggers to use Violation.

**Comment by AkMo3** [comment]
> > can we solve #9794 and #4465 in this single PR
> 
> it is better to do separately.
> it is ok in this PR to let Loggers to use Violation.

Done.

**Comment by romani** [comment]
the same 7 violations are in master code 
https://sonarcloud.io/project/issues?fileUuids=AW9t2waPYD2QG1pPXII7&id=org.checkstyle%3Acheckstyle&resolutions=WONTFIX

marked as WONTFIX
"it is API, it is hard to fix is now ... postponed till https://github.com/checkstyle/checkstyle/issues/7488"
"api/LocalizedMessage.java is API class .... we can not damage compatibility so easily. I hope we will find a way to deprecate such constructors."

**Review by romani — APPROVED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java [hunk 2]
```diff
@@ -78,8 +78,8 @@ public FileContents getFileContents() {
      *
      * @return the localized message
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java:79]: violation

File: src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java [hunk 3]
```diff
@@ -98,7 +98,7 @@ public int getLine() {
      * @return the event message
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java:98]: may be `violation message`?

File: src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java [hunk 4]
```diff
@@ -107,7 +107,7 @@ public String getMessage() {
      * @return the column associated with the message
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java:107]: violation

File: src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java [hunk 5]
```diff
@@ -116,7 +116,7 @@ public int getColumn() {
      * @return the column char index associated with the message
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java:116]: same

File: src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java [hunk 6]
```diff
@@ -135,7 +135,7 @@ public String getModuleId() {
      * @return the name of the source for the message
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java:135]: same

File: src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java [hunk 7]
```diff
@@ -144,7 +144,7 @@ public String getSourceName() {
      * @return the token type of the message
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java:144]: same

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java [hunk 8]
```diff
@@ -244,7 +244,7 @@ public final void log(int lineNo, int colNo, String key,
     }
 
     /**
-     * Helper method to log a LocalizedMessage.
+     * Helper method to log a Violation.
      *
      * @param ast a node to get line id column numbers associated
      *             with the message
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java:250]: with the violation

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java [hunk 9]
```diff
@@ -300,7 +300,7 @@ public final String getLine(int index) {
     private static class FileContext {
 
         /** The sorted set for collecting messages. */
-        private final SortedSet<LocalizedMessage> messages = new TreeSet<>();
+        private final SortedSet<Violation> messages = new TreeSet<>();
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java:303]: violations?

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java [hunk 10]
```diff
@@ -77,16 +77,16 @@ public void beginProcessing(String charset) {
     }
 
     @Override
-    public final SortedSet<LocalizedMessage> process(File file, FileText fileText)
+    public final SortedSet<Violation> process(File file, FileText fileText)
             throws CheckstyleException {
-        final SortedSet<LocalizedMessage> messages = context.get().messages;
+        final SortedSet<Violation> messages = context.get().messages;
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java:82]: violations?

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java [hunk 11]
```diff
@@ -112,11 +112,11 @@ protected final MessageDispatcher getMessageDispatcher() {
...
 
     /**
-     * Returns the sorted set of {@link LocalizedMessage}.
+     * Returns the sorted set of {@link Violation}.
      *
-     * @return the sorted set of {@link LocalizedMessage}.
+     * @return the sorted set of {@link Violation}.
      */
-    public SortedSet<LocalizedMessage> getMessages() {
+    public SortedSet<Violation> getMessages() {
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java:119]: getViolations?

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java [hunk 12]
```diff
@@ -191,18 +191,18 @@ public final void setTabWidth(int tabWidth) {
...
 
     /**
-     * Adds the sorted set of {@link LocalizedMessage} to the message collector.
+     * Adds the sorted set of {@link Violation} to the message collector.
      *
-     * @param messages the sorted set of {@link LocalizedMessage}.
+     * @param messages the sorted set of {@link Violation}.
      */
-    protected void addMessages(SortedSet<LocalizedMessage> messages) {
+    protected void addMessages(SortedSet<Violation> messages) {
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java:198]: addViolations?

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java [hunk 13]
```diff
@@ -248,7 +248,7 @@ protected final void fireErrors(String fileName) {
     private static class FileContext {
 
         /** The sorted set for collecting messages. */
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java:250]: violations?

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractViolationReporter.java [hunk 14]
```diff
@@ -24,7 +24,7 @@
 /**
  * Serves as an abstract base class for all modules that report inspection
  * findings. Such modules have a Severity level which is used for the
- * {@link LocalizedMessage localized messages} that are created by the module.
+ * {@link Violation localized messages} that are created by the module.
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractViolationReporter.java:27]: violations?

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AuditEvent.java [hunk 15]
```diff
@@ -46,7 +46,7 @@
     /** Filename event associated with. **/
     private final String fileName;
     /** Message associated with the event. **/
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AuditEvent.java:48]: Violation?

**Review by pbludov — CHANGES_REQUESTED** [review]
Please review all Javadoc and change "message" to "violation"

File: src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java (hunk 2 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java (hunk 3 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java (hunk 4 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java (hunk 5 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java (hunk 6 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java (hunk 7 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/TreeWalkerAuditEvent.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java (hunk 8 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java (hunk 9 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java (hunk 10 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java (hunk 11 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java (hunk 12 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java (hunk 13 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractFileSetCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractViolationReporter.java (hunk 14 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractViolationReporter.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AuditEvent.java (hunk 15 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/AuditEvent.java]: Done

**Comment by AkMo3** [comment]
> I think it is ok to NOT resolve such violations and #7488 in this PR. This update will be breaking compatibility, plugins and extensions would need be updated to compile Tests:
> https://github.com/jshiell/checkstyle-idea/search?q=LocalizedMessage (1 test)
> https://github.com/checkstyle/eclipse-cs/search?q=LocalizedMessage (0 items)
> https://github.com/checkstyle/sonar-checkstyle/search?q=LocalizedMessage (1 test)

I would send PR to each of these repositories for fix.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/FileSetCheck.java [hunk 16]
```diff
@@ -72,7 +72,7 @@
      * @return the sorted set of messages to be logged.
      * @throws CheckstyleException if error condition within Checkstyle occurs
      */
-    SortedSet<LocalizedMessage> process(File file, FileText fileText) throws CheckstyleException;
+    SortedSet<Violation> process(File file, FileText fileText) throws CheckstyleException;
```
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/FileSetCheck.java:75]: javadoc above says `The method should return the set of messages to be logged.` -> should be corrected

File: src/main/java/com/puppycrawl/tools/checkstyle/api/MessageDispatcher.java [hunk 17]
```diff
@@ -46,6 +46,6 @@
      * @param fileName the audited file
      * @param errors the audit errors from the file
```
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/MessageDispatcher.java:47]: not audit errors, but violations

File: src/main/java/com/puppycrawl/tools/checkstyle/api/Violation.java [hunk 18]
```diff
@@ -45,8 +45,8 @@
  *
  * @noinspection SerializableHasSerializationMethods, ClassWithTooManyConstructors
  */
-public final class LocalizedMessage
-    implements Comparable<LocalizedMessage>, Serializable {
+public final class Violation
```
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/Violation.java:48]: same here, javadoc should be corrected

File: src/xdocs/releasenotes.xml [hunk 19]
```diff
@@ -6333,7 +6333,7 @@ used in the MT mode.
             Boolean flag in Javadoc Checks that shows a current Javadoc comment has unclosed HTML tags. Author: Piyush Sharma <a href="https://github.com/checkstyle/checkstyle/issues/3311">#3311</a>
           </li>
           <li>
-            Add columnCharIndex field to LocalizedMessage. Author: Timur Tibeyev <a href="https://github.com/checkstyle/checkstyle/issues/4998">#4998</a>
+            Add columnCharIndex field to Violation. Author: Timur Tibeyev <a href="https://github.com/checkstyle/checkstyle/issues/4998">#4998</a>
```
Comment (strkkk) [inline src/xdocs/releasenotes.xml:6336]: I dont think it makes sense to change release notes, please revert

File: src/xdocs/releasenotes_old.xml [hunk 20]
```diff
@@ -2197,7 +2197,7 @@
         (allowThrowsTagsForSubclasses, request 540383)</li>
 
         <li>Added sourceName property to AuditEvent and
-        LocalizedMessage; and &quot;source&quot; attribute to
+        Violation; and &quot;source&quot; attribute to
```
Comment (strkkk) [inline src/xdocs/releasenotes_old.xml:2200]: same here

**Review by strkkk — CHANGES_REQUESTED** [review]

**Comment by strkkk** [comment]
Sonar is not happy https://sonarcloud.io/dashboard?id=org.checkstyle%3Acheckstyle&pullRequest=9761

File: src/main/java/com/puppycrawl/tools/checkstyle/api/FileSetCheck.java (hunk 16 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/FileSetCheck.java:75]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/api/MessageDispatcher.java (hunk 17 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/MessageDispatcher.java:47]: Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/Violation.java (hunk 18 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/Violation.java:48]: Done.

File: src/xdocs/releasenotes.xml (hunk 19 above)
Comment (AkMo3) [reply src/xdocs/releasenotes.xml]: Done.

File: src/xdocs/releasenotes_old.xml (hunk 20 above)
Comment (AkMo3) [reply src/xdocs/releasenotes_old.xml]: Done.

**Comment by romani** [comment]
@strkkk , here is explanation why sonar failures are ok https://github.com/checkstyle/checkstyle/pull/9761#issuecomment-813397729

**Comment by AkMo3** [comment]
@romani @rnveach @pbludov @strkkk I have some doubts regarding the renaming of methods. I have renamed some methods like [`getFilteredMessages -> getFilteredViolations`](https://github.com/AkMo3/checkstyle/blob/25796f93a48eba65d7758374b6c052f946bff207/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java#L180) but there are a lot more that can be renamed like [getMessage()](https://github.com/AkMo3/checkstyle/blob/25796f93a48eba65d7758374b6c052f946bff207/src/main/java/com/puppycrawl/tools/checkstyle/api/AuditEvent.java#L108) which can be changed to `getViolation()`, but I am afraid that this will lead to a lot more change in API then expected. Kindly share your views.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/MessageDispatcher.java (hunk 17 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/MessageDispatcher.java:47]: I do not see changes

Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/MessageDispatcher.java]: Please check again now.

**Review by strkkk — APPROVED** [review]

**Comment by strkkk** [comment]
> [``](https://github.com/AkMo3/checkstyle/blob/25796f93a48eba65d7758374b6c052f946bff207/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java#L180) but there are a lot more that can be renamed like [getMessage()](https://github.com/AkMo3/checkstyle/blob/25796f93a48eba65d7758374b6c052f946bff207/src/main/java/com/puppycrawl/tools/checkstyle/api/AuditEvent.java#L108) which can be changed to `getViolation()`

@AkMo3 I think it should be done in separate PR since current is too big already

**Comment by romani** [comment]
>I think it should be done in separate PR since current is too big already

Please create new issue on this and define what should be done

**Comment by AkMo3** [comment]
> @AkMo3, Please create new issue on this and define what should be done

Done at https://github.com/checkstyle/checkstyle/issues/9834

File: src/main/java/com/puppycrawl/tools/checkstyle/Checker.java [hunk 21]
```diff
@@ -320,8 +320,8 @@ private void processFiles(List<File> files) throws CheckstyleException {
      * @throws CheckstyleException if error condition within Checkstyle occurs.
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/Checker.java:320]: `@return a sorted set of messages to be logged.` => `@return a sorted set of violations to be logged.`

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractViolationReporter.java [hunk 22]
```diff
@@ -42,7 +42,7 @@
      *
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractViolationReporter.java:42]: `severity level of the messages` => `severity level of the violations`

**Review by pbludov — CHANGES_REQUESTED** [review]
Please fix couple of Javadocs

File: src/main/java/com/puppycrawl/tools/checkstyle/Checker.java (hunk 21 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/Checker.java:320]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractViolationReporter.java (hunk 22 above)
Comment (AkMo3) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractViolationReporter.java:42]: Done

**Comment by AkMo3** [comment]
Please restart cirrus JDK 
error log:
```
Failures
 - xmlstarlet - xmlstarlet not installed. The package was not found with the source(s) listed.
 Source(s): 'https://chocolatey.org/api/v2/'
 NOTE: When you specify explicit sources, it overrides default sources.
If the package version is a prerelease and you didn't specify `--pre`,
 the package may not be found.
Please see https://chocolatey.org/docs/troubleshooting for more 
 assistance.
 ```

**Review by pbludov — APPROVED** [review]
I'm OK to merge if CI passes.

**Comment by rnveach** [comment]
@romani How does this PR connect to https://github.com/checkstyle/checkstyle/issues/9794 . Will we merge this first and then work on other?

**Review by rnveach — APPROVED** [review]
I am ok if we are merging this.

**Comment by romani** [comment]
> Will we merge this first and then work on other?

yes.

**Comment by tsjensen** [comment]
Do you still maintain the **breaking** tag? Then this one should be tagged as breaking.

**Pull request merged**